src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs

```diff
@@ -219,11 +218,33 @@ private void ProcessSpans(
         }
     }
 
-    protected override bool TagEquals(IContainerStructureTag tag1, IContainerStructureTag tag2)
+    protected sealed override bool TagEquals(IContainerStructureTag latestTag, IContainerStructureTag previousTag)
     {
-        Contract.ThrowIfFalse(tag1 is StructureTag);
-        Contract.ThrowIfFalse(tag2 is StructureTag);
-        return tag1.Equals(tag2);
+        if (latestTag is not StructureTag latestStructureTag || previousTag is not StructureTag previousStructureTag)
+        {
+            Contract.Fail("Tags were the wrong type");
+            return latestTag == previousTag;
+        }
+
+        var latestSnapshot = latestStructureTag.Snapshot;
+        var previousSnapshot = previousStructureTag.Snapshot;
+
+        var previousStructureStart = new SnapshotPoint(previousSnapshot, previousStructureTag.HeaderSpan.Start);
```

Comment:
>HeaderSpan [](http://example.com/codeflow?start=94&length=10)

Is this just about the guideline behavior? If so, would GuideLineSpan.Start be appropriate to use? FYI, I believe you could also use a SnapshotSpan and SpanTrackingMode.Inclusive

Reply:
that seems reasonable.

Reply:
We can't use guidelinespan as we don't set it.  it's just:

```c#
 public Span? GuideLineSpan => null;
```

we defer to the editor to determine that span.

Reply:
I still feel like mapping a span is better than two calls to mapping points.

var previousStructureStart = new SnapshotSpan(previousSnapshot, previousStructureTag.HeaderSpan.Start, 0);

if (0 <previousStructureStart.TranslateTo(latestSnapshot, PointTrackingMode.Negative).Length)

Reply:
To clarify, I was thinking of something like the following (with the comment needing to be altered)

```
    protected sealed override bool TagEquals(IContainerStructureTag latestTag, IContainerStructureTag previousTag)
    {
        if (latestTag is not StructureTag latestStructureTag || previousTag is not StructureTag previousStructureTag)
        {
            Contract.Fail("Tags were the wrong type");
            return latestTag == previousTag;
        }

        // If these tags are the same object in memory, then they are obviously equal
        if (latestTag == previousTag)
            return true;

        var latestSnapshot = latestStructureTag.Snapshot;
        var previousSnapshot = previousStructureTag.Snapshot;

        var previousStructureStart = new SnapshotSpan(previousSnapshot, previousStructureTag.HeaderSpan.Start, 0);

        if (previousStructureStart.TranslateTo(latestSnapshot, SpanTrackingMode.EdgeInclusive).IsEmpty)
        {
            // We can't know that how we think this block moved is actually how the editor actually moved it.
            // Specifically, the tracking mode is an implementation detail.  As such, we don't want to reuse this tag as
            // its stale data (as mapped by the editor) may not be where we'd expect the new block's data to be.  This
            // can happen when the user types right at the start of a structure tag, causing it to move inwards
            // undesirably.

            return latestTag.Equals(previousTag);
        }

        return false;
    }
```